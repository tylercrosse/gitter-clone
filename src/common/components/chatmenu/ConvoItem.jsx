import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router';

const ConvoItem = ({ name, title }) => (
  <div className="convo-item">
    <Link className="convo-item-container" to={'/' + name}>
      <img className="convo-item-avatar" src={'http://i.pravatar.cc/22?u=' + name} alt={name} />
      <h2 className="convo-item-title">{title}</h2>
    </Link>
  </div>
);

ConvoItem.propTypes = {
  name: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
};

export default ConvoItem;
