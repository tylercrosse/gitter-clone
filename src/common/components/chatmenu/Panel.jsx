import React, { PropTypes } from 'react';
import CreateRoomModal      from '../modals/CreateRoomModal';
import ConvoItem            from './ConvoItem';

const Panel = (props) => {
  const active = props.active ? 'active' : '';
  return (
    <section
      onMouseLeave={props.onMouseLeave}
      // className="chat-menu-panel active"
      className={'chat-menu-panel ' + active}
    >
      <div className="brand-container">Gitter Clone</div>
      <header className="panel-header">
        <div className="panel-header-convos">
          <h2>All Conversations</h2>
        </div>
      </header>
      <div className="panel-inner">
        <footer className="panel-footer">
          <div className="panel-footer-convos">
            <button
              onClick={props.openCreateRoomModal}
              className="create-rooom-button"
            >
              Add a room
            </button>
            <CreateRoomModal
              user={props.user}
              modalIsOpen={props.modalIsOpen}
              onRequestClose={props.closeModal}
              onFormSubmit={props.addConvo}
            />
          </div>
        </footer>
        <div className="panel-content scroller">
          <section className="primary-collection">
            {props.convos && Object.values(props.convos).map((convo) =>
              <ConvoItem
                key={convo.id}
                {...convo}
              />
            )}
          </section>
        </div>
      </div>
    </section>
  );
};

Panel.propTypes = {
  onMouseLeave: PropTypes.func.isRequired,
  openCreateRoomModal: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  addConvo: PropTypes.func.isRequired,
  modalIsOpen: PropTypes.bool.isRequired,
  user: PropTypes.object.isRequired
};

export default Panel;
