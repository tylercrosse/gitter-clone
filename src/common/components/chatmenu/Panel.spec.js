import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Panel from './Panel';
import ConvoItem from './ConvoItem';

const setup = () => {
  const props = {
    modalIsOpen: false,
    onMouseLeave: jest.fn(),
    openCreateRoomModal: jest.fn(),
    closeModal: jest.fn(),
    addConvo: jest.fn(),
    fetchConvos: jest.fn(),
    user: {
      loggedIn: true
    },
    convos: {
      '5898b14f6047fd8db8b4e3ba': {
        id: '5898b14f6047fd8db8b4e3ba',
        updatedAt: '2017-02-06T17:24:31.593Z',
        createdAt: '2017-02-06T17:24:31.593Z',
        name: 'chat',
        __v: 0
      },
      '5898b15d6047fd8db8b4e3bb': {
        id: '5898b15d6047fd8db8b4e3bb',
        updatedAt: '2017-02-06T17:24:45.657Z',
        createdAt: '2017-02-06T17:24:45.657Z',
        name: 'chat2',
        __v: 0
      }
    }
  };

  const component = <Panel {...props} />;

  return {
    props,
    component
  };
};

describe('<Panel />', () => {
  it('should render correctly', () => {
    const { component } = setup();
    const renderedComponent = renderer.create(component);
    const tree = renderedComponent.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a <ConvoItem /> for every convo', () => {
    const { props, component } = setup();
    const wrapper = shallow(component);
    expect(wrapper.find(ConvoItem).length)
      .toEqual(Object.keys(props.convos).length);
  });

  it('should toggle active state onMouseLeave', () => {
    const { component, props } = setup();
    const wrapper = mount(component);
    expect(props.onMouseLeave).not.toHaveBeenCalled();
    wrapper.find('.chat-menu-panel').simulate('mouseLeave');
    expect(props.onMouseLeave).toHaveBeenCalled();
  });

  it('should apply class "active" when props.active == true', () => {
    const { component, props } = setup();
    const wrapper = mount(component);
    expect(wrapper.props().active).toBeFalsy();
    expect(wrapper.find('.chat-menu-panel').hasClass('active'))
      .toBeFalsy();
    wrapper.setProps({...props, active: true});
    expect(wrapper.props().active).toBeTruthy();
    expect(wrapper.find('.chat-menu-panel').hasClass('active'))
      .toBeTruthy();
  });
});
