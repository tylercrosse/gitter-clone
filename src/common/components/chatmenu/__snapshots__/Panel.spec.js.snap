// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Panel /> should render correctly 1`] = `
<section
  className="chat-menu-panel "
  onMouseLeave={[Function]}
>
  <div
    className="brand-container"
  >
    Gitter Clone
  </div>
  <header
    className="panel-header"
  >
    <div
      className="panel-header-convos"
    >
      <h2>
        
      </h2>
    </div>
  </header>
</section>
`;
